### PR TITLE
fix(docs-infra): convert hard-coded `comparing-observables` examples into a proper mini-app

### LINKED ISSUE #31024: docs: observables examples hard-coded and outdated
# 📚 Docs or angular.io bug report

### Description

The [Comparing Observables][1] guide shows examples with the old, obsolete syntax of having operators on the `Observable` prototype, instead of using `.pipe()`.

Also, the examples are hard-coded, instead of being extracted from actual mini-apps, which would prevent the code from getting out-of-date or break.

### What's the affected URL?**
https://angular.io/guide/comparing-observables

[1]: https://angular.io/guide/comparing-observables

### PR DESCRIPTION
**Pull request by sonukapoor**
…stable files

It also updates some indention issues with the code that is part of inline tables.

Fixes #31024

## PR Checklist
Please check if your PR fulfills the following requirements:

- [X] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [X] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [X] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:

## What is the current behavior?

Issue Number: #31024

## Does this PR introduce a breaking change?

- [ ] Yes
- [X] No

### CODE REVIEW

File: aio/content/examples/comparing-observables/e2e/src/app.e2e-spec.ts [hunk 1]
```diff
@@ -0,0 +1,9 @@
+/*
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/e2e/src/app.e2e-spec.ts]: Can you please remove this file from this PR and add it to #34190 (once this one is merged)?
It would be better to have all the test infrastructure code in one PR. Let's keep this PR about getting rid of the hard-coded examples.

File: aio/content/examples/comparing-observables/example-config.json [hunk 2]
```diff
@@ -0,0 +1,8 @@
+{
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/example-config.json]: Can you please remove this file from this PR and add it to #34190 (once this one is merged)?
It would be better to have all the test infrastructure code in one PR. Let's keep this PR about getting rid of the hard-coded examples.

File: aio/content/guide/comparing-observables.md [hunk 3]
```diff
@@ -165,7 +159,6 @@ subscription.unsubscribe();</pre>
 <pre>function handler(e) {
   console.log(‘Clicked’, e);
 }
-
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: This change seems unrelated. Please revert.

File: aio/content/guide/comparing-observables.md [hunk 4]
```diff
@@ -100,14 +90,16 @@ The following code snippets illustrate how the same kind of operation is defined
     <tr>
       <td>Creation</td>
       <td>
-        <pre>new Observable((observer) => {
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: :100:

For reference, here is how this change looks like:

**Before:**
observable-cheatsheet before

**After:**
observable-cheatsheet after

File: aio/content/guide/comparing-observables.md [hunk 5]
```diff
@@ -21,67 +21,57 @@ Observables are often compared to promises. Here are some key differences:
...
 * Observable subscriptions are cancellable. Unsubscribing removes the listener from receiving further values, and notifies the subscriber function to cancel work.
 
-<code-example hideCopy>
-const sub = obs.subscribe(...);
-sub.unsubscribe();
+<code-example path="comparing-observables/src/observables.ts" header="comparing-observables/src/observables.ts" region="unsubscribe">
+
 </code-example>
 
+
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: Why this extra line?

File: aio/content/guide/comparing-observables.md [hunk 6]
```diff
@@ -21,67 +21,57 @@ Observables are often compared to promises. Here are some key differences:
...
 
-<code-example hideCopy>
-// declare a publishing operation
-new Observable((observer) => { subscriber_fn });
-// initiate execution
-observable.subscribe(() => {
-  // observer handles notifications
-});
+<code-example path="comparing-observables/src/observables.ts" header="comparing-observables/src/observables.ts" region="observables">
+
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: Now that there is not content in these `<code-example>` elements, can you put the closing tag on the same line (or format them with one attribute per line for readability:

```html
<!-- Either: -->
<code-example path="comparing-observables/src/observables.ts" region="observables" header="comparing-observables/src/observables.ts"></code-example>

<!-- Or: -->
<code-example
    path="comparing-observables/src/observables.ts"
    region="observables"
    header="comparing-observables/src/observables.ts">
</code-example>
```

(Same for all other occurrences below.)

File: aio/content/guide/comparing-observables.md [hunk 7]
```diff
@@ -21,67 +21,57 @@ Observables are often compared to promises. Here are some key differences:
...
 * Observables are not executed until a consumer subscribes. The `subscribe()` executes the defined behavior once, and it can be called again. Each subscription has its own computation. Resubscription causes recomputation of values.
 
-<code-example hideCopy>
-// declare a publishing operation
-new Observable((observer) => { subscriber_fn });
-// initiate execution
-observable.subscribe(() => {
-  // observer handles notifications
-});
+<code-example path="comparing-observables/src/observables.ts" header="comparing-observables/src/observables.ts" region="observables">
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: The `header` should be as short as possible and its purpose is to give the user context about which file in a hypothetical app. So, no need to include the app directory:

```diff
-header="comparing-observables/src/observables.ts" region="observables"
+header="src/observables.ts" region="observables"
```

(Same for all other occurrences below.)

File: aio/content/examples/comparing-observables/src/observables.ts [hunk 8]
```diff
@@ -0,0 +1,25 @@
+
+import { Observable } from 'rxjs';
+
+// #docregion observables
+
+// declare a publishing operation
+const observable = new Observable((observer) => {
+  // ...
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables.ts]: To keep it more similar to the original snippet:

```diff
-  // ...
+  // Subscriber fn...
```

File: aio/content/examples/comparing-observables/src/promise-error.ts [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+const promise = new Promise((resolve, reject) => {
+  // ...
+});
+
+// #docregion
+// handle return value
+promise.then((value) => {
+  throw Error('my error');
+});
+// #enddocregion
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise-error.ts]: Please, merge this with the `promise.ts` file. There is not reason to have two different files (especially given that they have common code).

File: aio/content/examples/comparing-observables/src/promise.ts [hunk 10]
```diff
@@ -0,0 +1,14 @@
+// Added export {} to get rid of the 'error TS2451: Cannot redeclare block-scoped variable 'promise'.'
+export {};
+
+// #docregion
+// initiate execution
+const promise = new Promise((resolve, reject) => {
+  // ...
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise.ts]: To keep it more similar to the original snippet:

```diff
-  // ...
+  // Executer fn...
```

File: aio/content/examples/comparing-observables/src/observables.ts [hunk 11]
```diff
@@ -0,0 +1,25 @@
+
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables.ts]: Unnecessary empty line.

File: aio/content/examples/comparing-observables/src/observables-chaining.ts [hunk 12]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables-chaining.ts]: Unnecessary empty line.

File: aio/content/examples/comparing-observables/src/observables-chaining.ts [hunk 13]
```diff
@@ -0,0 +1,14 @@
...
+// declare a publishing operation
+const observable = new Observable((observer) => {
+  // ...
+});
+
+// #docregion observable-chain
+
+observable.pipe(map((v: any) => 2 * v));
+
+// #enddocregion
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables-chaining.ts]: Please, merge this with the `observables.ts` file. There is not reason to have two different files (especially given that they have common code).

File: aio/content/examples/comparing-observables/src/promise.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+// Added export {} to get rid of the 'error TS2451: Cannot redeclare block-scoped variable 'promise'.'
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise.ts]: Let's rename this file to `promises.ts` to be consistent with `observables.ts`.

File: aio/content/examples/comparing-observables/src/promise-chaining.ts [hunk 15]
```diff
@@ -0,0 +1,13 @@
...
+// initiate execution
+const promise = new Promise((resolve, reject) => {
+  // ...
+});
+
+// #docregion promise-chain
+
+promise.then((v: any) => 2 * v);
+
+// #enddocregion
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise-chaining.ts]: Please, merge this with the `promise.ts` file. There is not reason to have two different files (especially given that they have common code).

File: aio/content/examples/comparing-observables/src/observables-chaining.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+import { Observable } from 'rxjs';
+
+// declare a publishing operation
+const observable = new Observable((observer) => {
+  // ...
+});
+
+// #docregion observable-chain
+
+observable.pipe(map((v: any) => 2 * v));
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables-chaining.ts]: I believe it is better to move the type to the `Observable` instantiation. E.g.: `const observable = new Observable<number>(...)`

File: aio/content/examples/comparing-observables/src/promise-chaining.ts [hunk 17]
```diff
@@ -0,0 +1,13 @@
...
+export {};
+
+// initiate execution
+const promise = new Promise((resolve, reject) => {
+  // ...
+});
+
+// #docregion promise-chain
+
+promise.then((v: any) => 2 * v);
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise-chaining.ts]: I believe it is better to move the type to the `Promise` instantiation. E.g.: `const promise = new Promise<number>(...)`

File: aio/content/examples/comparing-observables/src/observables.ts [hunk 18]
```diff
@@ -0,0 +1,25 @@
...
+observable.subscribe(() => {
+  // observer handles notifications
+});
+
+// #enddocregion observables
+
+// #docregion unsubscribe
+const subscription = observable.subscribe(() => {
+  // observer handles notifications
+});
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables.ts]: I think you can share this block with both docregions with something like:

```
// #docregion observables

// declare a publishing operation
const observable = new Observable((observer) => {
  // ...
});

// initiate execution

// #docregion unsubscribe
const subscription = observable.subscribe(() => {
  // observer handles notifications
});

// #enddocregion observables
subscription.unsubscribe();

// #enddocregion unsubscribe
```

File: aio/content/examples/comparing-observables/src/observables-error.ts [hunk 19]
```diff
@@ -0,0 +1,14 @@
...
+  // ...
+});
+
+// #docregion
+
+observable.subscribe(() => {
+  throw Error('my error');
+});
+
+// #enddocregion
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/observables-error.ts]: Please, merge this with the `observables.ts` file. There is not reason to have two different files (especially given that they have common code).

File: aio/content/examples/comparing-observables/src/promise-error.ts [hunk 20]
```diff
@@ -0,0 +1,14 @@
+// Added export {} to get rid of the 'error TS2451: Cannot redeclare block-scoped variable 'promise'.'
+export {};
+
+// initiate execution
+const promise = new Promise((resolve, reject) => {
+  // ...
+});
+
+// #docregion
+// handle return value
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise-error.ts]: This comment seems unrelated. Please remove.

File: aio/content/examples/comparing-observables/src/promise-error.ts [hunk 21]
```diff
@@ -0,0 +1,14 @@
...
+promise.then((value) => {
```
Comment (gkalpak) [inline aio/content/examples/comparing-observables/src/promise-error.ts]:
```diff
-promise.then((value) => {
+promise.then(() => {
```

File: aio/content/guide/comparing-observables.md (hunk 7 above)
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: Can you, also, try indenting these `<code-example>` elements to see if our markdown parser is able to detect them as part of the bullet point?

**Review by gkalpak — CHANGES_REQUESTED** [review]
Off to a great start :tada:
Some more fine-tuning and this should be good to go! 

(BTW, why the empty `main.ts` file?)

File: aio/content/guide/comparing-observables.md (hunk 4 above)
Comment (sonukapoor) [reply aio/content/guide/comparing-observables.md]: @gkalpak: I am assuming you are happy with this? The brackets are inlined properly with the beginning of the first character. See below:

image

Comment (gkalpak) [reply aio/content/guide/comparing-observables.md]: Yes, the :100: emoji means [I am very happy][1] :grin:

[1]: https://emojipedia.org/hundred-points-symbol

File: aio/content/guide/comparing-observables.md [hunk 22]
```diff
@@ -21,47 +21,50 @@ Observables are often compared to promises. Here are some key differences:
...
-<code-example hideCopy>
-// declare a publishing operation
-new Observable((observer) => { subscriber_fn });
-// initiate execution
-observable.subscribe(() => {
-  // observer handles notifications
-});
+<code-example 
+  path="comparing-observables/src/observables.ts" 
+  header="src/observables.ts (observables)"
```
Comment (gkalpak) [inline aio/content/guide/comparing-observables.md]: Let's call this `observable` (probably makes sense to rename the region too), to be consistent with `promise` below.

**Review by gkalpak — CHANGES_REQUESTED** [review]
Shaping up nicely ✨
I left a few more comments (nothing major).

File: aio/angular.json [hunk 23]
```diff
@@ -189,4 +190,4 @@
     }
   },
   "defaultProject": "site"
-}
+}
```
[the remark this answers is not in the document]
Comment (sonukapoor) [reply aio/angular.json]: I am not sure why this keeps happening, but whenever I change branches and run `yarn start`, it asks me if I want to enable `analytics`. Regardless, of my response, it modifies the `angular.json` file, which I undo before the commit. I will revert it but wanted to let you know why this change was there.

[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/angular.json]: Strange. Do you have `@angular/cli` installed globally?

[the remark this answers is not in the document]
Comment (sonukapoor) [reply aio/angular.json]: Yes, I do.

File: aio/content/guide/comparing-observables.md [hunk 24]
```diff
@@ -165,7 +172,6 @@ subscription.unsubscribe();</pre>
 <pre>function handler(e) {
   console.log(‘Clicked’, e);
 }
-
```
[the remark this answers is not in the document]
Comment (sonukapoor) [reply aio/content/guide/comparing-observables.md]: For some reason, the font changes, when there is a line-break in between. See below the font-size after the line-break:

image

File: aio/content/guide/comparing-observables.md (hunk 22 above)
Comment (sonukapoor) [reply aio/content/guide/comparing-observables.md]: Sounds good. Going to rename `promises.ts` to `promise.ts` as well.

File: aio/content/guide/comparing-observables.md [hunk 25]
```diff
@@ -21,47 +21,50 @@ Observables are often compared to promises. Here are some key differences:
...
 * Observables are not executed until a consumer subscribes. The `subscribe()` executes the defined behavior once, and it can be called again. Each subscription has its own computation. Resubscription causes recomputation of values.
 
-<code-example hideCopy>
-// declare a publishing operation
-new Observable((observer) => { subscriber_fn });
-// initiate execution
-observable.subscribe(() => {
-  // observer handles notifications
-});
+<code-example
```
[the remark this answers is not in the document]
Comment (sonukapoor) [reply aio/content/guide/comparing-observables.md]: Done.

File: aio/content/guide/comparing-observables.md (hunk 22 above)
Comment (gkalpak) [reply aio/content/guide/comparing-observables.md]: I was actually referring to the docregion :grin:
The file names were fine.

File: aio/content/guide/comparing-observables.md (hunk 24 above)
[the remark this answers is not in the document]
Comment (gkalpak) [reply aio/content/guide/comparing-observables.md]: I see. That's some quirk of our Markdown parser. Not worth looking into it.
Let's leave it as is. Thx for looking into it :+1:

**Review by gkalpak — CHANGES_REQUESTED** [review]
Almost there :tada:

Can you please take care of https://github.com/angular/angular/pull/34327#discussion_r357566813 (rename the files to plural and the docregion to singular), squash the commits into one and update the commit message to something along the lines of dd9793f92?

File: aio/content/guide/comparing-observables.md (hunk 22 above)
Comment (sonukapoor) [reply aio/content/guide/comparing-observables.md]: No problem. I think I got it this time :)

**Comment by sonukapoor** [comment]
@gkalpak I believe this resolves everything. LMK if anything is missing.

**Review by gkalpak — APPROVED** [review]
Great work, @sonukapoor! Thx 💯

**Comment by sonukapoor** [comment]
@IgorMinar Can you please review this?

**Review by Splaktar — APPROVED** [review]
LGTM

**Review by IgorMinar — APPROVED** [review]
Looks great! Thank you

**Comment by gkalpak** [comment]
@caretaker: Global approval for the docs-only changes.
